artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/connect/mirror/AMQPMirrorControllerTarget.java

```diff
@@ -177,6 +180,48 @@ public void onError(int errorCode, String errorMessage) {
 
    private AckManager ackManager;
 
+   /** This method will wait both replication and storage to finish their current operations. */
+   public void flush() {
+      CountDownLatch latch = new CountDownLatch(1);
+      connection.runNow(() -> {
+         OperationContext oldContext = OperationContextImpl.getContext();
+         try {
+            OperationContextImpl.setContext(mirrorContext);
+            mirrorContext.executeOnCompletion(new IOCallback() {
+               @Override
+               public void done() {
+                  latch.countDown();
+               }
+
+               @Override
+               public void onError(int errorCode, String errorMessage) {
+                  logger.warn("error code = {} / message = {}", errorCode, errorMessage);
+                  latch.countDown();
+               }
+            });
+         } finally {
+            OperationContextImpl.setContext(oldContext);
+         }
+      });
+
+      long timeout;
+      try {
+         timeout = connection.getProtocolManager().getAckManagerFlushTimeout();
+      } catch (Throwable e) {
+         logger.warn(e.getMessage(), e);
+         timeout = 10_000;
+      }
```

Comment:
This one could do with even having a message of its own. The potential failure would seem to be an NPE if the connection or protocol manager isnt there, in which case its just going to depend on what JVM is being used whether it says anything useful at all (newer JVMs will indicate what was null...old ones wont).

Reply:
I thought of potential issues in the integration-test suite... but even that.. it's theoretical. 

it's redundant code that is not meant to ever happen.